<launch>
    <arg name="ns" default="/demo"/>
     

    <group ns="$(arg ns)/sim">
        <node pkg="turtlesim" name="turtlesim" type="turtlesim_node"/>
        <node pkg="turtlesine" name="mapsim" type="mapsim"/>

    </group>

    <group ns="$(arg ns)">    
        <arg name="turtlename" default="turtle1"/>
        <node pkg="turtlesine" name="$(arg turtlename)_driver" type="turtlesine">
           
            <param name="a_speed" value="4.44" type="double" />
            <param name="l_speed" value="4.44" type="double" />
            
            <param name="initial_x" value="5.5" type="double" />
            <param name="initial_y" value="5.7" type="double" />
            <param name="initial_theta" value="1" type="double" />

            <remap from="~pose" to="$(arg ns)/sim/$(arg turtlename)/pose" />
            <remap from="~cmd_vel" to="$(arg ns)/sim/$(arg turtlename)/cmd_vel" />
            <remap from="~spawn" to="$(arg ns)/sim/spawn" />
            <remap from="~kill" to="$(arg ns)/sim/kill" />
            <remap from="~teleport_absolute" to="$(arg ns)/sim/$(arg turtlename)/teleport_absolute" />   
        </node>  
        <node pkg="turtlesine" name="$(arg turtlename)_laser1" type="lasersim">

            <param name="measurments" value="600" type="int" />
            <param name="angle_min" value="-0.52359" type="double" />
            <param name="angle_max" value="0.52359" type="double" />
            <param name="range_min" value="0.05" type="double" />
            <param name="range_max" value="3.0" type="double" />
            <param name="turtles_cnt" value="2" type="int" />
            <param name="turtletarget" value="$(arg ns)/turtletarget" type="str" />

            <remap from="~walls" to="$(arg ns)/sim/mapsim/walls" />
            <remap from="~stampedpose2" to="$(arg ns)/turtle2_driver/stampedpose" />
        </node>  

        <node pkg="turtlesine" name="$(arg turtlename)_laser2" type="lasersim">  
            <param name="measurments" value="600" type="int" />
            <param name="angle_min" value="-0.52359" type="double" />
            <param name="angle_max" value="0.52359" type="double" />
            <param name="range_min" value="0.05" type="double" />
            <param name="range_max" value="3.0" type="double" />
            <param name="turtles_cnt" value="2" type="int" />
            <param name="turtletarget" value="$(arg ns)/turtletarget" type="str" />

            <remap from="~walls" to="$(arg ns)/sim/mapsim/walls" />
            <remap from="~stampedpose2" to="$(arg ns)/turtle2_driver/stampedpose" />
        </node>

        <node pkg="turtlesine" name="$(arg turtlename)_laser3" type="lasersim">
            <param name="measurments" value="600" type="int" />
            <param name="angle_min" value="-0.52359" type="double" />
            <param name="angle_max" value="0.52359" type="double" />
            <param name="range_min" value="0.05" type="double" />
            <param name="range_max" value="3.0" type="double" />
            <param name="turtles_cnt" value="2" type="int" />
            <param name="turtletarget" value="$(arg ns)/turtletarget" type="str" />

            <remap from="~walls" to="$(arg ns)/sim/mapsim/walls" />
            <remap from="~stampedpose2" to="$(arg ns)/turtle2_driver/stampedpose" />
        </node>

        <node pkg="turtlesine" name="$(arg turtlename)_laser4" type="lasersim"> 
            <param name="measurments" value="600" type="int" />
            <param name="angle_min" value="-0.52359" type="double" />
            <param name="angle_max" value="0.52359" type="double" />
            <param name="range_min" value="0.05" type="double" />
            <param name="range_max" value="3.0" type="double" />
            <param name="turtles_cnt" value="2" type="int" />
            <param name="turtletarget" value="$(arg ns)/turtletarget" type="str" />

            <remap from="~walls" to="$(arg ns)/sim/mapsim/walls" />
            <remap from="~stampedpose2" to="$(arg ns)/turtle2_driver/stampedpose" />
        </node>
        
        <node pkg="tf" type="static_transform_publisher" name="$(arg turtlename)_to_laser1" args=" 0.3  0.2 0  0.78    0 0 $(arg turtlename)_base_link $(arg turtlename)_laser1 100" />  
        <node pkg="tf" type="static_transform_publisher" name="$(arg turtlename)_to_laser2" args=" 0.3 -0.2 0 -0.78    0 0 $(arg turtlename)_base_link $(arg turtlename)_laser2 100" /> 

        <node pkg="tf" type="static_transform_publisher" name="$(arg turtlename)_to_laser3" args="-0.3  0.2 0  2.35    0 0 $(arg turtlename)_base_link $(arg turtlename)_laser3 100" />  
        <node pkg="tf" type="static_transform_publisher" name="$(arg turtlename)_to_laser4" args="-0.3 -0.2 0 -2.35    0 0 $(arg turtlename)_base_link $(arg turtlename)_laser4 100" /> 
       
    </group>

    <group ns="$(arg ns)">    
        <arg name="turtlename" default="turtle2"/>
    
        <node pkg="turtlesine" name="$(arg turtlename)_driver" type="turtlesine">

            <param name="follow_frame" value="turtle1_base_link" type="str" />
            <param name="turtletarget" value="$(arg ns)/turtletarget" type="str" />
            
            <param name="a_speed" value="1.44" type="double" />
            <param name="l_speed" value="1.44" type="double" />
            
            <param name="initial_x" value="11.08" type="double" />
            <param name="initial_y" value="5.5444" type="double" />
            <param name="initial_theta" value="4.8" type="double" />

            <remap from="~pose" to="$(arg ns)/sim/$(arg turtlename)/pose" />
            <remap from="~cmd_vel" to="$(arg ns)/sim/$(arg turtlename)/cmd_vel" />
            <remap from="~spawn" to="$(arg ns)/sim/spawn" />
            <remap from="~kill" to="$(arg ns)/sim/kill" />
            <remap from="~teleport_absolute" to="$(arg ns)/sim/$(arg turtlename)/teleport_absolute" />
            
        </node>  
        <node pkg="turtlesine" name="$(arg turtlename)_laser1" type="lasersim">

            <param name="measurments" value="628" type="int" />
            <param name="angle_min" value="-0.52359" type="double" />
            <param name="angle_max" value="0.52359" type="double" />
            <param name="range_min" value="0.05" type="double" />
            <param name="range_max" value="6.0" type="double" />
            <param name="turtles_cnt" value="2" type="int" />

            <remap from="~walls" to="$(arg ns)/sim/mapsim/walls" />
            <remap from="~stampedpose1" to="$(arg ns)/turtle1_driver/stampedpose" />          
            
        </node>
        <node pkg="turtlesine" name="$(arg turtlename)_laser2" type="lasersim">

            <param name="measurments" value="628" type="int" />
            <param name="angle_min" value="-0.52359" type="double" />
            <param name="angle_max" value="0.52359" type="double" />
            <param name="range_min" value="0.05" type="double" />
            <param name="range_max" value="6.0" type="double" />
            <param name="turtles_cnt" value="2" type="int" />

            <remap from="~walls" to="$(arg ns)/sim/mapsim/walls" />
            <remap from="~stampedpose1" to="$(arg ns)/turtle1_driver/stampedpose" />          
            
        </node>  
        <node pkg="turtlesine" name="$(arg turtlename)_laser3" type="lasersim">

            <param name="measurments" value="628" type="int" />
            <param name="angle_min" value="-0.52359" type="double" />
            <param name="angle_max" value="0.52359" type="double" />
            <param name="range_min" value="0.05" type="double" />
            <param name="range_max" value="6.0" type="double" />
            <param name="turtles_cnt" value="2" type="int" />

            <remap from="~walls" to="$(arg ns)/sim/mapsim/walls" />
            <remap from="~stampedpose1" to="$(arg ns)/turtle1_driver/stampedpose" />          
            
        </node>  
        <node pkg="turtlesine" name="$(arg turtlename)_laser4" type="lasersim">

            <param name="measurments" value="628" type="int" />
            <param name="angle_min" value="-0.52359" type="double" />
            <param name="angle_max" value="0.52359" type="double" />
            <param name="range_min" value="0.05" type="double" />
            <param name="range_max" value="6.0" type="double" />
            <param name="turtles_cnt" value="2" type="int" />

            <remap from="~walls" to="$(arg ns)/sim/mapsim/walls" />
            <remap from="~stampedpose1" to="$(arg ns)/turtle1_driver/stampedpose" />          
            
        </node>  
        <node pkg="tf" type="static_transform_publisher" name="$(arg turtlename)_to_laser1" args=" 0.3  0.2 0  0.78    0 0 $(arg turtlename)_base_link $(arg turtlename)_laser1 100" />  
        <node pkg="tf" type="static_transform_publisher" name="$(arg turtlename)_to_laser2" args=" 0.3 -0.2 0 -0.78    0 0 $(arg turtlename)_base_link $(arg turtlename)_laser2 100" /> 

        <node pkg="tf" type="static_transform_publisher" name="$(arg turtlename)_to_laser3" args="-0.3  0.2 0  2.35    0 0 $(arg turtlename)_base_link $(arg turtlename)_laser3 100" />  
        <node pkg="tf" type="static_transform_publisher" name="$(arg turtlename)_to_laser4" args="-0.3 -0.2 0 -2.35    0 0 $(arg turtlename)_base_link $(arg turtlename)_laser4 100" />   
    </group>

</launch>